include/bamit/IntervalNode.hpp

```diff
@@ -199,80 +199,78 @@ void construct_tree(std::unique_ptr<IntervalNode> & node, std::vector<Record> co
-        {
-            break;
-        }
-        current_reads.emplace_back(std::make_tuple(r.reference_id().value(), r.reference_position().value()),
-                                   std::make_tuple(r.reference_id().value(), r.reference_position().value() +
-                                                                             get_length(r.cigar_sequence())),
-                                   r.file_offset());
-    }
+    // If the current median is overlapping the read, take the current offset as result and search the left further.
     if (cur_median >= start && cur_median <= end)
     {
-        results.insert(std::end(results), std::begin(current_reads), std::end(current_reads));
-        overlap(input, root->get_left_node(), start, end, results);
-        overlap(input, root->get_right_node(), start, end, results);
+        offstream_pos = root->get_file_offset();
+        overlap(root->get_left_node(), start, end, offstream_pos);
     }
-    // If current median is to the right of the overlap, sort reads in ascending order and add all reads which
-    // start before the overlap ends.
+    // If current median is to the right of the overlap, go to the left tree.
     else if (end < cur_median)
     {
-        std::sort(current_reads.begin(), current_reads.end(), RecordComparatorStart());
-        for (auto const & r : current_reads)
-        {
-            if (r.start <= end)
-            {
-                results.push_back(r);
-            }
-            else
-            {
-                break;
-            }
-        }
-        overlap(input, root->get_left_node(), start, end, results);
+        offstream_pos = root->get_file_offset();
+        overlap(root->get_left_node(), start, end, offstream_pos);
     }
+    // If current median is to the left of the overlap, go to the right tree.
     else if (start > cur_median)
     {
-        std::sort(current_reads.begin(), current_reads.end(), RecordComparatorEnd());
-        for (auto const & r : current_reads)
+        offstream_pos = root->get_file_offset();
+        overlap(root->get_right_node(), start, end, offstream_pos);
+    }
+}
+
+/*!
+   \brief Find the records which overlap a given start and end position.
+   \param input The sam file input of type bamit::sam_file_input_type.
+   \param root The root of the tree to search in.
+   \param start The start position of the search.
+   \param end The end position of the search.
+   \param results The list of records overlapping the search.
+*/
+void get_records(sam_file_input_type & input,
+                 std::unique_ptr<IntervalNode> const & root,
+                 Position const & start,
+                 Position const & end,
+                 std::filesystem::path & outname)
+{
+    std::streamoff offstream_pos{-1};
+    overlap(root, start, end, offstream_pos);
+
+    input.seek(offstream_pos);
+
+    seqan3::sam_file_output fout{outname, bamit::sam_file_output_fields{}};
```

Comment:
I'm not sure how to do this, but it would be nice if this can switch between output to `outname` vs output to `std::cout` based on whether `outname.empty()` is true or not...

Reply:
I would think about this in a later PR, when we have decided how we want to print the records?